### PR TITLE
TN-1975 always preserve order of QBT.at values inserting QB_Timeline

### PR DESCRIPTION
**Pull request by pbugni**
New data structure added to maintain order of list by attribute `at` field, named `AtOrderedList`.
The `append` method on `AtOrderedList` enforces order on insertion.
Use when building user's qb_timeline rows to prevent order errors, such as those demonstrated in TN-1975.

NB - no migration yet to seek out any qb_timeline rows needing to be reprocessed.  Waiting to see if another issue's migration during same sprint will cover.

### CODE REVIEW

File: portal/models/qb_timeline.py [hunk 1]
```diff
@@ -62,6 +62,50 @@ def qbd(self):
...
+class AtOrderedList(list):
+    """Specialize ``list`` to maintain insertion order and ``at`` values
+
+    When building up QBTs for a user, we need to maintain both order and
+    sort by ``QBT.at`` values.  As there are rows with identical ``at``
+    values, it can't simply be sorted by a field.
+
+    As the build order matters, continue to add QBTs to the end of the list,
+    taking special care to insert those with earlier ``at`` values in
+    the correct place.  Two or more identicle ``at`` values should result
```
Comment (ivan-c) [inline portal/models/qb_timeline.py]:
```suggestion
    the correct place.  Two or more identical ``at`` values should result
```

File: portal/models/qb_timeline.py [hunk 2]
```diff
@@ -62,6 +62,50 @@ def qbd(self):
             recur_id=self.qb_recur_id, qb_id=self.qb_id)
 
 
+class AtOrderedList(list):
+    """Specialize ``list`` to maintain insertion order and ``at`` values
```
Comment (ivan-c) [inline portal/models/qb_timeline.py]: `at values` seems too jargony. Something between "QB timeline snapshot creation datetime" and `at` would be ideal

File: tests/test_qb_timeline.py [hunk 3]
```diff
@@ -420,3 +421,89 @@ def test_age(self):
...
+
+
+class HasAt(object):
+    """Mock the sorted portion of QBT for testing AtOrderedList"""
+    def __init__(self, at, label):
+        self.at = at
+        self.label = label
+
+
+n = HasAt(datetime.utcnow(), 'now')
```
Comment (ivan-c) [inline tests/test_qb_timeline.py]: Please pick an arbitrary datetime, [non-deterministic tests can be flakey](https://martinfowler.com/articles/nonDeterminism.html#Time)

File: tests/test_qb_timeline.py [hunk 4]
```diff
@@ -420,3 +421,89 @@ def test_age(self):
...
+y = HasAt(n.at - relativedelta(days=1), 'yesterday')
+y2 = HasAt(y.at, 'yesterday again')
+wa = HasAt(n.at - relativedelta(weeks=1), 'week ago')
```
Comment (ivan-c) [inline tests/test_qb_timeline.py]: Please use longer variable names. Unless they get long enough to cause PEP8 nags, longer variable names are easier to grep/find/replace

File: portal/models/qb_timeline.py [hunk 5]
```diff
@@ -370,7 +414,7 @@ def update_users_QBT(user_id, invalidate_existing=False):
 
         # As we move forward, capture state at each time point
 
-        pending_qbts = []
+        pending_qbts = AtOrderedList()
```
Comment (ivan-c) [inline portal/models/qb_timeline.py]: Why not just define a sorted "view" (function) of the list (eg `.keys()`, `.items()`, `.values()`)?

Comment (pbugni) [reply portal/models/qb_timeline.py]: tried to make this clear in the comment - we need to preserve insertion order AND then by `at`.  Using a built in sorted view won't retain this need.

**Comment by pbugni** [comment]
@ivan-c said:
> Had some code style nags, ready to merge otherwise

Committed changes to address.  Please review AYC

**Pull request merged**